### ISSUE TITLE
[inductor] Attention-shaped pattern rewrite ignores the softmax dim, silently computing attention over the last axis

### ISSUE THREAD

**Issue by laolvfan** [pytorch/pytorch#196468]
### 🐛 Describe the bug

An attention-shaped chain whose softmax is taken over a **non-last** dimension is
rewritten by the post-grad pattern matcher as if the softmax were over the last
dimension. The compiled function silently returns the result of a *different*
computation — no error, no warning, no graph break.

```python
import torch, torch.nn as nn

class M(nn.Module):
    def __init__(self, sm_dim):
        super().__init__()
        self.dropout = nn.Dropout(0.23)
        self.sm_dim = sm_dim
    def forward(self, q, k, v):
        qk = torch.matmul(q, k.transpose(-2, -1))
        s = qk.div(128)
        s = torch.softmax(s, dim=self.sm_dim)   # dim=0, NOT the last dim
        s = self.dropout(s)
        return s.matmul(v)

torch.manual_seed(0)
q, k, v = torch.randn(4,4,2,8), torch.randn(4,4,3,8), torch.randn(4,4,3,8)

def ev(m):
    m.train(False)
    with torch.no_grad():
        return m(q, k, v)

def cp(m):
    m.train(False); torch._dynamo.reset()
    with torch.no_grad():
        return torch.compile(m, backend="inductor")(q, k, v)

rel = lambda a, b: ((a-b).abs().max() / a.abs().max()).item()

print(rel(ev(M(0)),  cp(M(0))))    # 0.3462      eager vs compiled, dim=0
print(rel(cp(M(0)),  ev(M(-1))))   # 6.5e-08     compiled(dim=0) == eager(dim=-1)
```

### The compiled `dim=0` model computes the `dim=-1` model

The second number is the point: the compiled output of the `dim=0` model is
numerically identical to the **eager output of the `dim=-1` model**, so the
rewrite has silently replaced the user's softmax axis with the last axis.

### It is the pattern matcher

Disabling the pattern matcher removes the divergence entirely, and a model that
really does use `dim=-1` is unaffected:

| configuration | max relative diff (eager vs compiled) |
|---|---|
| `softmax(dim=0)`, `pattern_matcher=True` | **8.72** (shapes `(50,50,4,64)`) / **0.346** (shapes `(4,4,2,8)`) |
| `softmax(dim=0)`, `pattern_matcher=False` | 7.0e-08 |
| `softmax(dim=-1)`, `pattern_matcher=True` | 1.7e-07 |

```python
import torch._inductor.config as cfg
cfg.pattern_matcher = False   # divergence disappears
```

`nn.Dropout` is in eval mode throughout (`m.train(False)`), so it is an identity
and the comparison is deterministic — the eager output is bit-stable over 200
repeated runs.

### Note on related reports

This is a different root cause from #195320 (which is about a permuted `K` being
bound as if it were `key.transpose(-2,-1)`). Here `k.transpose(-2, -1)` is the
canonical form and the operands are bound correctly; the only mismatched
attribute is the softmax dimension, which the rewrite does not check.

### Versions

Reproduced on `2.14.0+cpu`, CPU, inductor backend.

cc @ezyang @gchanan @kadeng @msaroufim @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @muchulee8 @amjames @aakhundov @coconutruben @jataylo @drisspg @liangel-02 @howardzhang-cv

**Comment by zlaabsi** [comment]
> AI-assisted local verification (Codex), posted at the contributor's request.
>
> #195383 already covers this case. I verified its diff at `e0963aed691053681f4e3ef72a135a41539a8a1e` against this issue's CPU reproduction, including the dropout-in-eval configuration and seed 0.
>
> | Softmax dim | Relative error before | With #195383 | SDPA fusions after |
> | --- | ---: | ---: | ---: |
> | `0` | 0.3461925 | 0 | 0 |
> | `-2` | 0.3352560 | 0 | 0 |
> | `-1` | 6.51e-8 | 6.51e-8 | 1 |
> | `3` | 6.51e-8 | 6.51e-8 | 1 |
>
> Disabling `pattern_matcher` also removes the divergence on the unpatched baseline. The existing PR rejects the non-last-axis cases while preserving both equivalent spellings of the valid last-axis fusion.
>
> Environment: macOS arm64 / M1 Max, CPU Inductor, official `2.15.0.dev20260910` native binaries with their matching Python source base `d8ae8d5990dc2a34d044f313d788ae0dc3fbb668`. The table compares that baseline with the diff of #195383 applied in a separate checkout.
>
> Its two CPU `non_last_dim_softmax` regression tests also pass, for static and dynamic shapes. On macOS I invoked the imported test classes explicitly because the file's script entry point is Linux-only:
>
> ```sh
> PYTHONPATH=. python -c 'from test.inductor.test_fused_attention import *; run_tests()' -k non_last_dim_softmax
> # Ran 2 tests; OK
> ```
>
> This is CPU verification of the existing PR, not a new fix or a full-suite/GPU validation.

**Comment by zlaabsi** [comment]
> #195383 already covers this case. I verified its diff at `e0963aed691053681f4e3ef72a135a41539a8a1e` against this issue's CPU reproduction, including the dropout-in-eval configuration and seed 0.
>
> | Softmax dim | Relative error before | With #195383 | SDPA fusions after |
> | --- | ---: | ---: | ---: |
> | `0` | 0.3461925 | 0 | 0 |
> | `-2` | 0.3352560 | 0 | 0 |
> | `-1` | 6.51e-8 | 6.51e-8 | 1 |
> | `3` | 6.51e-8 | 6.51e-8 | 1 |
>
> Disabling `pattern_matcher` also removes the divergence on the unpatched baseline. The existing PR rejects the non-last-axis cases while preserving both equivalent spellings of the valid last-axis fusion.
>
> Environment: macOS arm64 / M1 Max, CPU Inductor, official `2.15.0.dev20260910` native binaries with their matching Python source base `d8ae8d5990dc2a34d044f313d788ae0dc3fbb668`. The table compares that baseline with the diff of #195383 applied in a separate checkout.
>
> Its two CPU `non_last_dim_softmax` regression tests also pass, for static and dynamic shapes. On macOS I invoked the imported test classes explicitly because the file's script entry point is Linux-only:
>
> ```sh
> PYTHONPATH=. python -c 'from test.inductor.test_fused_attention import *; run_tests()' -k non_last_dim_softmax
> # Ran 2 tests; OK
> ```
>
> This is CPU verification of the existing PR, not a new fix or a full-suite/GPU validation.

Codex was used.

**Comment by laolvfan** [comment]
Cross-reference: #195383 appears to address the same root cause as this report.

That PR revalidates serialized pattern matches against a pattern retraced with the actual matched inputs, and uses canonical dimension matching for the `amax`/`sum` dimensions that implement softmax. It also adds `_test_pattern_fails_with_non_last_dim_softmax`, which verifies that a non-last-axis softmax is declined while an equivalent positive spelling of the last axis still fuses.

The regression test uses `dim=-2` rather than this issue's `dim=0`, but both exercise the same previously wildcarded softmax-dimension attribute. I will retest this issue's original reproducer once #195383 lands.
